File: scripts/reference-solutions/spiral-matrix.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int m, n;
  cin >> m >> n;
  vector<vector<long long>> grid(m, vector<long long>(n));
  for (int r = 0; r < m; ++r) {
    for (int c = 0; c < n; ++c) {
      cin >> grid[r][c];
    }
  }

  int top = 0, bottom = m - 1, left = 0, right = n - 1;
  bool first = true;
  while (top <= bottom && left <= right) {
    for (int c = left; c <= right; ++c) {
      if (!first) cout << ' ';
      first = false;
      cout << grid[top][c];
    }
    ++top;
    for (int r = top; r <= bottom; ++r) {
      if (!first) cout << ' ';
      first = false;
      cout << grid[r][right];
    }
    --right;
    if (top <= bottom) {
      for (int c = right; c >= left; --c) {
        if (!first) cout << ' ';
        first = false;
        cout << grid[bottom][c];
      }
      --bottom;
    }
    if (left <= right) {
      for (int r = bottom; r >= top; --r) {
        if (!first) cout << ' ';
        first = false;
        cout << grid[r][left];
      }
      ++left;
    }
  }
  cout << '\n';
  return 0;
}

File: scripts/reference-solutions/spiral-matrix.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    m = int(data[0])
    n = int(data[1])
    vals = list(map(int, data[2 : 2 + m * n]))
    grid = [vals[r * n : (r + 1) * n] for r in range(m)]

    top, bottom, left, right = 0, m - 1, 0, n - 1
    res = []
    while top <= bottom and left <= right:
        for c in range(left, right + 1):
            res.append(grid[top][c])
        top += 1
        for r in range(top, bottom + 1):
            res.append(grid[r][right])
        right -= 1
        if top <= bottom:
            for c in range(right, left - 1, -1):
                res.append(grid[bottom][c])
            bottom -= 1
        if left <= right:
            for r in range(bottom, top - 1, -1):
                res.append(grid[r][left])
            left += 1

    sys.stdout.write(" ".join(map(str, res)) + "\n")


main()
